File: typescript/src/namespaces/warden.ts
/**
 * Warden namespace — safety checks, intent verification, and consensus.
 */

export class WardenNamespace {
  /** @internal */
  constructor(private callDg: (tool: string, params: Record<string, any>) => Promise<any>,
              private warn: (method: string) => void) {}

  /** Run a canary safety check (`data-grout/warden.canary`). */
  async canary(params: Record<string, any>): Promise<any> {
    this.warn('warden.canary');
    return this.callDg('warden.canary', params);
  }

  /** Verify intent before executing an action (`data-grout/warden.intent`). */
  async verifyIntent(params: Record<string, any>): Promise<any> {
    this.warn('warden.intent');
    return this.callDg('warden.intent', params);
  }

  /** Adjudicate a dispute or ambiguity (`data-grout/warden.adjudicate`). */
  async adjudicate(params: Record<string, any>): Promise<any> {
    this.warn('warden.adjudicate');
    return this.callDg('warden.adjudicate', params);
  }

  /** Multi-model ensemble consensus check (`data-grout/warden.ensemble`). */
  async ensemble(params: Record<string, any>): Promise<any> {
    this.warn('warden.ensemble');
    return this.callDg('warden.ensemble', params);
  }
}

File: rust/src/namespaces/warden.rs
//! Warden namespace — safety checks, intent verification, and consensus.

use crate::error::Result;
use crate::Client;
use serde_json::Value;

/// Safety gates, intent verification, and multi-model consensus.
///
/// Obtained via [`Client::warden`].
pub struct Warden<'a>(pub(crate) &'a Client);

impl<'a> Warden<'a> {
    /// Run a canary safety check (`data-grout/warden.canary`).
    pub async fn canary(&self, params: Value) -> Result<Value> {
        self.0.warn_if_not_dg("warden.canary");
        self.0.ensure_initialized().await?;
        self.0
            .call_dg_tool("data-grout/warden.canary", params)
            .await
    }

    /// Verify intent before executing an action (`data-grout/warden.intent`).
    pub async fn verify_intent(&self, params: Value) -> Result<Value> {
        self.0.warn_if_not_dg("warden.intent");
        self.0.ensure_initialized().await?;
        self.0
            .call_dg_tool("data-grout/warden.intent", params)
            .await
    }

    /// Adjudicate a dispute or ambiguity (`data-grout/warden.adjudicate`).
    pub async fn adjudicate(&self, params: Value) -> Result<Value> {
        self.0.warn_if_not_dg("warden.adjudicate");
        self.0.ensure_initialized().await?;
        self.0
            .call_dg_tool("data-grout/warden.adjudicate", params)
            .await
    }

    /// Multi-model ensemble consensus check (`data-grout/warden.ensemble`).
    pub async fn ensemble(&self, params: Value) -> Result<Value> {
        self.0.warn_if_not_dg("warden.ensemble");
        self.0.ensure_initialized().await?;
        self.0
            .call_dg_tool("data-grout/warden.ensemble", params)
            .await
    }
}
